File: src/Ferrovie/Massa.java
package Ferrovie;

import java.util.Objects;

public abstract class Massa implements Comparable<Massa>{
    private double m;
    public Massa(double m){
        if(m<0.0 || m>255.00){
            throw new IllegalArgumentException("Massa in Tonnellate non compresa [0 ; 255]");
        }
        this.m = m;
    }

    public Massa(){}

    public double getMassa(){ return this.m;}

    @Override
    public String toString(){
        return "Massa: " + this.getMassa();
    }

    @Override
    public boolean equals(Object o){
        if(this == o){return true;}
        if(o == null){return false;}
        if(o instanceof Massa c){
            return this.getMassa() == c.getMassa();
        }
        return false;
    }

    @Override
    public int hashCode(){
        return Objects.hash(this.m);
    }

    @Override
    public int compareTo(Massa o) {
        return Double.compare(this.m, o.getMassa());
    }

}

File: Python_Esercizi/Ferrovia/Massa.py
class Massa:
    def __init__(self , m):
        if(m < 0.0 or m > 255.0):
            raise ValueError("Massa non valide, non compresa tra 0 e 255")
        self._m = m

    def __init__(self):
        self._ro = 0

    def getMassa(self):
        return self._m

    def __eq__(self, other):
        if(other is None): return False
        if(self is other): return True
        if(isinstance(other, Massa)):
            return self.getMassa() == other.getMassa()
        return False

    def __ne__(self, other):
        return not self == other

    def __str__(self):
        return "Massa: " + str(self.getMassa()) + "t"

    def __gt__(self, other):
        if(other is not None and isinstance(other, Massa)):
            return self.getMassa() > other.getMassa()
        raise TypeError("Massa non valide")

    def __lt__(self , other):
        if(other is not None and isinstance(other, Massa)):
            return self.getMassa() < other.getMassa()
        raise TypeError("Massa non valide")

    def __ge__(self , other):
        if(other == self):
            return True
        return self > other

    def __le__(self , other):
        if(other == self):
            return True
        return self < other

    def __hash__(self):
        return hash(self.getMassa())

